substrate/frame/sassafras/src/lib.rs

```diff
@@ -0,0 +1,1062 @@
+	/// tickets list.
+	///
+	/// If `slot` value falls within the next epoch then we fetch tickets from the next epoch
+	/// tickets ids list. Note that in this case we may have not finished receiving all the tickets
+	/// for that epoch yet. The next epoch tickets should be considered "stable" only after the
+	/// current epoch first half slots were elapsed (see `submit_tickets_unsigned_extrinsic`).
+	///
+	/// Returns `None` if, according to the sorting strategy, there is no ticket associated to the
+	/// specified slot-index (happend if a ticket falls in the middle of an epoch and n > k),
+	/// or if the slot falls beyond the next epoch.
+	///
+	/// Before importing the first block this returns `None`.
+	pub fn slot_ticket_id(slot: Slot) -> Option<TicketId> {
+		if frame_system::Pallet::<T>::block_number() < One::one() {
+			return None
+		}
+		let epoch_idx = EpochIndex::<T>::get();
+		let epoch_len = T::EpochLength::get();
+		let mut slot_idx = Self::slot_index(slot);
+		let mut tickets_meta = TicketsMeta::<T>::get();
+
+		let get_ticket_idx = |slot_idx| {
+			let ticket_idx = if slot_idx < epoch_len / 2 {
+				2 * slot_idx + 1
+			} else {
+				2 * (epoch_len - (slot_idx + 1))
+			};
+			debug!(
+				target: LOG_TARGET,
+				"slot-idx {} <-> ticket-idx {}",
+				slot_idx,
+				ticket_idx
+			);
+			ticket_idx as u32
+		};
+
+		let mut epoch_tag = (epoch_idx & 1) as u8;
+
+		if epoch_len <= slot_idx && slot_idx < 2 * epoch_len {
+			// Try to get a ticket for the next epoch. Since its state values were not enacted yet,
+			// we may have to finish sorting the tickets.
+			epoch_tag ^= 1;
+			slot_idx -= epoch_len;
+			if tickets_meta.unsorted_tickets_count != 0 {
+				Self::sort_tickets(u32::MAX, epoch_tag, &mut tickets_meta);
+				TicketsMeta::<T>::set(tickets_meta);
+			}
+		} else if slot_idx >= 2 * epoch_len {
+			return None
+		}
+
+		let ticket_idx = get_ticket_idx(slot_idx);
+		if ticket_idx < tickets_meta.tickets_count[epoch_tag as usize] {
+			TicketsIds::<T>::get((epoch_tag, ticket_idx))
+		} else {
+			None
+		}
+	}
+
+	/// Returns ticket id and data associated to the given `slot`.
+	///
+	/// Refer to the `slot_ticket_id` documentation for the slot-ticket association
+	/// criteria.
+	pub fn slot_ticket(slot: Slot) -> Option<(TicketId, TicketBody)> {
+		Self::slot_ticket_id(slot).and_then(|id| TicketsData::<T>::get(id).map(|body| (id, body)))
+	}
+
+	// Lexicographically sort the tickets which belong to the next epoch.
+	//
+	// Tickets are fetched from at most `max_segments` segments.
+	//
+	// The resulting sorted vector is optionally truncated to contain at most `MaxTickets`
+	// entries. If all the unsorted segments are consumed then the sorted vector is
+	// saved as the next epoch tickets, else it is saved to be used by next calls to
+	// this function.
+	pub(crate) fn sort_tickets(max_segments: u32, epoch_tag: u8, metadata: &mut TicketsMetadata) {
```

Comment:
As far as I can see doesn't this has an invariant that there are no duplicate tickets IDs? If so can you document this and maybe add some `debug_assert`s that verify it (I don't know, maybe by e.g. having a `BTreeSet` with all IDs if `#[cfg(debug_assertions)` is true)?

Reply:
`TicketId` is a u128 VRF output so there should be no collisions. That said we should probably make this obvious in the documentation. This invariant is assumed in the `TicketsData` storage map, since it's using `Identity` as hash function (maybe that's the place where we could put a comment).

Reply:
> `TicketId` is a u128 VRF output so there should be no collisions.

Yes, but there's nothing stopping someone from submitting the same ticket multiple times, right? And if this isn't filtered out before this function is called it can break.

Reply:
Not sure we are talking about the same thing but tickets duplicated are dropped

Indeed you spotted the bug where the `continue` was missing here:

https://github.com/paritytech/polkadot-sdk/blob/f45ef2f290767577f5ef85a00500373500103c00/substrate/frame/sassafras/src/lib.rs#L443-L446

Every ticket which is accepted as valid is then inserted in the `TicketsData` map which ensures the uniqueness, regardless of submission time and epoch (the is only one map)